lib/ansible/plugins/connection/nspawn.py

```diff
@@ -0,0 +1,152 @@
+
+        if os.geteuid() != 0:
+            raise AnsibleError("nspawn connection requires running as root")
+
+        # we're running as root on the local system so do some
+        # trivial checks for ensuring 'host' is actually a chroot'able dir
+        if not os.path.isdir(self.chroot):
+            raise AnsibleError("%s is not a directory" % self.chroot)
+
+        chrootsh = os.path.join(self.chroot, 'bin/sh')
+        if not is_executable(chrootsh):
+            raise AnsibleError("%s does not look like a chrootable dir (/bin/sh missing)" % self.chroot)
+
+        self.nspawn_cmd = 'systemd-nspawn'
+
+    def _connect(self):
+      pass
+
+    def _buffered_exec_command(self, cmd, stdin=subprocess.PIPE):
+        ''' run a command on the chroot.  This is only needed for
+        implementing put_file() get_file() so that we don't have to
+        read the whole file into memory.
+
+        compared to exec_command() it looses some niceties like being
+        able to return the process's exit code immediately.
+        '''
+        executable = (
+                C.DEFAULT_EXECUTABLE.split()[0]
+                if C.DEFAULT_EXECUTABLE
+                else '/bin/sh')
+
+        nspawn_args = shlex.split(self._play_context.nspawn_args)
+        local_cmd = [self.nspawn_cmd, '-D', self.chroot ] + nspawn_args + [
+                '--', executable, '-c', cmd]
+
+        display.vvv("EXEC %s" % (local_cmd), host=self.chroot)
+        local_cmd = map(to_bytes, local_cmd)
+        p = subprocess.Popen(local_cmd, shell=False, stdin=stdin,
+                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        return p
+
+    def exec_command(self, cmd, in_data=None, sudoable=False):
+        ''' run a command on the chroot '''
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        p = self._buffered_exec_command(cmd)
+        stdout, stderr = p.communicate(in_data)
+        return (p.returncode, stdout, stderr)
+
+    def _prefix_login_path(self, remote_path):
+        ''' Make sure that we put files into a standard path
+
+            If a path is relative, then we need to choose where to put it.
+            ssh chooses $HOME but we aren't guaranteed that a home dir will
+            exist in any given chroot.  So for now we're choosing "/" instead.
+            This also happens to be the former default.
+
+            Can revisit using $HOME instead if it's a problem
+        '''
+        if not remote_path.startswith(os.path.sep):
+            remote_path = os.path.join(os.path.sep, remote_path)
+        return os.path.normpath(remote_path)
+
+    def put_file(self, in_path, out_path):
+        ''' transfer a file from local to chroot '''
+        super(Connection, self).put_file(in_path, out_path)
+        display.vvv("PUT %s TO %s" % (in_path, out_path), host=self.chroot)
+
+        out_path = pipes.quote(self._prefix_login_path(out_path))
```

Comment:
We've switched from pipes.quote()  to six.moves.shlex._quote for python3 compatibility:
``` python
 from ansible.compat.six.moves import shlex_quote
 out_path = shlex_quote(self._prefix_login_path(out_path))
```